src/Symfony/Component/Lock/Store/AbstractStore.php

```diff
@@ -0,0 +1,30 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock\Store;
+
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\Lock;
+use Symfony\Component\Lock\StoreInterface;
+
+/**
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+abstract class AbstractStore implements StoreInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function createLock($resource, $ttl = 300.0)
+    {
+        return new Lock(new Key($resource), $this, $ttl);
```

Comment:
That was the first implementation (I like the separation of responsibility too).
But, as suggested by @nicolas-grekas in [this comment](https://github.com/symfony/symfony/pull/21093#pullrequestreview-14675328) I move the factory into the StoreInterface and that's why I create the AbstractStore to avoid code duplication (which I don't really like too). 

The issue with the factory is was an empty class like without real value:
```php
class StoreFactory
{
    private $store;

    public function __construct(StoreInterface $store)
    {
        $this->store = $store;
    }

    public function createLock($resource, $ttl = 300.0)
    {
        return new Lock(new Key($resource), $this->lock, $ttl);
    }
}
```

And IMO, the outer store may create the lock instead of delegating the creation to the inner store. The abstractStore do the job (maybe a trait could be a better solution than the abstraction) 

What do you think ?

(just to be sure, you are talking about decoration instead of composition ?)